File: min_stack/MinStack.java
class MinStack {
    private Deque<int[]> stack;
    
    public MinStack() {
        this.stack = new ArrayDeque<>();
    }
    
    public void push(int val) {
        int min = this.stack.isEmpty() ? val : Math.min(val, this.stack.peekFirst()[1]);
        this.stack.offerFirst(new int[]{val, min});
    }
    
    public void pop() {
        if (this.stack.isEmpty()) {
            return;
        }
        this.stack.pollFirst();
    }
    
    public int top() {
        if (this.stack.isEmpty()) {
            return Integer.MIN_VALUE;
        }
        return this.stack.peekFirst()[0];
    }
    
    public int getMin() {
        if (this.stack.isEmpty()) {
            return Integer.MIN_VALUE;
        }
        return this.stack.peekFirst()[1];
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: min_stack/Min_Stack.cpp
class Min_Stack {
private:
    stack<int*>* min_stack;
    
public:
    MinStack() {
        this->min_stack = new stack<int*>();
    }
    
    void push(int val) {
        int curr_min = (*min_stack).empty() ? val : min((*min_stack).top()[1], val);
        (*min_stack).push(new int[]{val, curr_min});
    }
    
    // c++中pop后自动调用指针的析构函数。返回类型是void。这点和java不同。
    void pop() {
        if ((*min_stack).empty()) {
            return;
        }
        (*min_stack).pop();
    }
    
    int top() {
        if ((*min_stack).empty()) {
            return INT_MIN;
        }
        return (*min_stack).top()[0];
    }
    
    int getMin() {
        if ((*min_stack).empty()) {
            return INT_MIN;
        }
        return (*min_stack).top()[1]; 
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */